from_images: python-native czi support and timepoint and channel multiplexing

jstriebel opened the pull request:
### Description:
- `Dataset.from_images` now adds a layer per timepoint and per channel (if the data doesn't have 1 or 3 channels).
- Added python-native CZI support for `Dataset.from_images` or `dataset.add_layer_from_images`, without using bioformats.
- `dataset.add_layer_from_images` can add a layer per timepoint and per channel when passing `allow_multiple_layers=True`.
- Fixes for the pims ImageIO reader are improved.

### Issues:
- part of #748

### Todos:
 - [x] Updated Changelog
 - [x] Updated Documentation
 - [x] Added / Updated Tests

jstriebel commented on webknossos/webknossos/dataset/dataset.py:
```diff
@@ -987,7 +989,10 @@ def add_layer_from_images(
         use_bioformats: bool = False,
         channel: Optional[int] = None,
         timepoint: Optional[int] = None,
+        czi_channel: Optional[int] = None,
```
I have to admit that this whole method is growing quite large. It might be a good idea to extract it into its own module, but not quite sure.

philippotto commented on webknossos/webknossos/dataset/_utils/pims_images.py:
```diff
@@ -178,10 +200,23 @@ def __init__(
                     else:
                         self._img_dims = "yxc"
                     self._iter_dim = "z"
+                elif len(images.shape) == 5:
+                    if images.shape[2] == 1 or (
+                        images.shape[2] == 3 and images.dtype == np.dtype("uint8")
+                    ):
+                        self._img_dims = "cyx"
+                    else:
+                        self._img_dims = "yxc"
+                    self._iter_dim = "z"
+                    self._timepoint = 0
+                    if images.shape[0] > 1:
+                        self._possible_layers["timepoint"] = list(
+                            range(0, images.shape[0])
+                        )
```
some comments would be nice. jumping from another domain into this code is quite hard :see_no_evil:

philippotto commented on webknossos/webknossos/dataset/_utils/pims_czi_reader.py:
```diff
@@ -0,0 +1,112 @@
+    def class_exts(cls) -> Set[str]:
+        return {".czi"}
+
+    class_priority = 20
```
what is this?

jstriebel replied:
This is used in pims to pick the reader with the highest priority. Default is 10, and bioformats priority (which is the only other reader supporting czi) is 2. From the [pims docs](http://soft-matter.github.io/pims/v0.6.1/custom_readers.html#plugging-into-pims-s-open-function):
> To prioritize readers, a field class_priority can optionally be given to the reader. A higher priority will be chosen over a lower priority. Default for all readers is 10.

jstriebel replied:
Also added as a comment.

philippotto commented on webknossos/webknossos/dataset/_utils/pims_czi_reader.py:
```diff
@@ -0,0 +1,112 @@
+        plane.pop("X", None)
+        plane.pop("Y", None)
+        plane["C"] = self.czi_channel
+        print(plane)
```
remove

philippotto commented on webknossos/webknossos/dataset/_utils/pims_czi_reader.py:
```diff
@@ -0,0 +1,112 @@
+
+    def get_frame_2D(self, **ind: int) -> np.ndarray:
+        plane = {k.upper(): v for k, v in ind.items()}
+        plane.pop("X", None)
+        plane.pop("Y", None)
```
why are these popped?

philippotto reviewed (changes requested):
From a higher level this PR looks good, but it could use lots of code comments, because it's hard to grasp the domain concepts when one isn't really deep in this topic.

> Dataset.from_images now adds a layer per timepoint and per channel (if the data doesn't have 1 or 3 channels).

Is there a layer limit? Will this create 100 layers when there are 100 timepoints?

jstriebel commented on webknossos/tests/dataset/test_add_layer_from_images.py:
```diff
@@ -220,7 +251,7 @@ def test_bioformats(
         "https://samples.scif.io/test-jpeg2000.zip",
         "scifio-test.jp2",
-        {},
+        {},  # {"channel": 499},
```
Sorry, that was a left-over, removed it.

jstriebel commented on webknossos/webknossos/dataset/_utils/pims_images.py:
```diff
@@ -113,13 +127,20 @@ def __init__(
+                    # https://github.com/soft-matter/pims/pull/430
                     if (
-                        len(images.shape) > len(images.sizes)
+                        len(images._shape) >= len(images.sizes)
```
Yes, `_shape` has a dimension less than `shape`^^ That part is quite confusing atm, I hope that the actual fix in pims will be merged.

jstriebel commented on webknossos/webknossos/dataset/dataset.py:
```diff
@@ -1032,103 +1039,158 @@ def add_layer_from_images(
             use_bioformats=use_bioformats,
             is_segmentation=category == "segmentation",
         )
+        possible_layers = pims_images.get_possible_layers()
+        if (
+            possible_layers is not None
+            and truncate_rgba_to_rgb
+            and len(possible_layers.get("channel", [])) == 4
+        ):
+            if len(possible_layers) == 1:
+                possible_layers = None
+            else:
+                del possible_layers["channel"]
+        if possible_layers is not None:
+            if allow_multiple_layers:
+                suffix_with_kwargs = {
+                    "__" + "_".join(f"{k}={v}" for k, v in sorted(pairs)): dict(pairs)
+                    for pairs in product(
+                        *(
+                            [(key, value) for value in values]
+                            for key, values in possible_layers.items()
+                        )
+                    )
+                }
+                print(suffix_with_kwargs)
+            else:
+                suffix_with_kwargs = {"": {}}
+                warnings.warn(
+                    f"There are dimensions beyond channels and xyz which cannot be read: {possible_layers}. "
+                    "Defaulting to the first one. "
+                    "Please set allow_multiple_layers=True if all of them should be written to different layers, "
+                    "or set specific values as arguments.",
+                    RuntimeWarning,
+                )
+        else:
+            suffix_with_kwargs = {"": {}}
```
> (suffix doesn't seem to be used by the way).

`suffix` is used as a layer name suffix below in `self.add_layer(…)`, I renamed it to `layer_name_suffix`. Also added many more comments what's happening there and renamed some vars to make it more clear.

jstriebel commented:
@philippotto Thanks a lot for the review!

> From a higher level this PR looks good, but it could use lots of code comments, because it's hard to grasp the domain concepts when one isn't really deep in this topic.

I added a couple comments, refactored some parts slightly and used better variable names. What do you think about it now? Where is still more explanation necessary?

> > Dataset.from_images now adds a layer per timepoint and per channel (if the data doesn't have 1 or 3 channels).
> 
> Is there a layer limit? Will this create 100 layers when there are 100 timepoints?

There's not limit currently. `ds.add_layer_from_images` by default only adds a single one, `allow_multiple_layers` must be set to `True` explicitly, but `Dataset.from_images` automatically would add any number of layers right now. Do you think we should set an upper limit? I'd guess only for `Dataset.from_images` then, which might also be circumvented manually?

philippotto reviewed (commented):
Thank you for improving the code readability :) There's one code location left where I had some trouble with understanding. Other than that, it's already looking pretty good :)

jstriebel commented on webknossos/webknossos/dataset/_utils/pims_images.py:
```diff
@@ -170,40 +174,53 @@ def __init__(
+                    if timepoint is not None:
+                        raise RuntimeError(
+                            f"Got {len(images.shape)} axes for the images, "
+                            + "cannot map to 3D+channels+timepoints."
```
Ah, this error message is misleading. This only is thrown if timepoint is already set, and the image is 5d after removing the time dimension. I improved the error message.

jstriebel commented on webknossos/webknossos/dataset/dataset.py:
```diff
@@ -1032,103 +1040,168 @@ def add_layer_from_images(
+            # and automatically truncate to 3 channels
+            # (pims_images takes care of this:)
+            del possible_layers["channel"]
+        # Below, we iterate over suffix_with_pims_open_kwargs_per_layer in the for-loop
+        # to add one layer per possible_layer if allow_multiple_layers is True.
+        # If just a single layer is added, we still add a default value in the dict.
```
Not directly, changed it to `Further below, …`

jstriebel commented on webknossos/webknossos/dataset/dataset.py:
```diff
@@ -1032,103 +1040,168 @@ def add_layer_from_images(
+        # If just a single layer is added, we still add a default value in the dict.
+        if possible_layers is not None and len(possible_layers) > 0:
+            if allow_multiple_layers:
+                suffix_with_pims_open_kwargs_per_layer = {
+                    "__" + "_".join(f"{k}={v}" for k, v in sorted(pairs)): dict(pairs)
+                    for pairs in product(
+                        *(
+                            [(key, value) for value in values]
+                            for key, values in possible_layers.items()
+                        )
+                    )
+                }
```
Good idea, I added an example :+1:

jstriebel commented on webknossos/webknossos/dataset/dataset.py:
```diff
@@ -1032,103 +1040,168 @@ def add_layer_from_images(
+                    RuntimeWarning,
+                )
+        else:
+            suffix_with_pims_open_kwargs_per_layer = {"": {}}
```
Exactly :+1: Also added a comment about it.

jstriebel commented:
@philippotto Thanks for re-checking, I added some more comments :writing_hand: What do you think about the number of layers, should we limit it to some maximum?

> > Is there a layer limit? Will this create 100 layers when there are 100 timepoints?
> 
> There's not limit currently. `ds.add_layer_from_images` by default only adds a single one, `allow_multiple_layers` must be set to `True` explicitly, but `Dataset.from_images` automatically would add any number of layers right now. Do you think we should set an upper limit? I'd guess only for `Dataset.from_images` then, which might also be circumvented manually?

philippotto commented:
> What do you think about the number of layers, should we limit it to some maximum?
> 
> > >

I think more than 20 layers are not really feasible to render in webKnossos (at the moment). This is why I'd suggest to limit it so that only the first X timepoints are used. There could be an "ignore_layer_limit" parameter if the user wants to circumvent this. But I don't think it's necessary.

jstriebel commented:
> I think more than 20 layers are not really feasible to render in webKnossos (at the moment). This is why I'd suggest to limit it so that only the first X timepoints are used. There could be an "ignore_layer_limit" parameter if the user wants to circumvent this. But I don't think it's necessary.

I added a `max_layers` keyword-only argument with 20 as the default. Is there anything else missing from your side?

philippotto reviewed (approved):
Awesome :)
